tests/cycle/main.rs

```diff
@@ -0,0 +1,755 @@
+    MaxPanic(Inputs),
+
+    /// value of the given input, plus one
+    Successor(Box<Input>),
+}
+
+impl Input {
+    fn eval(self, db: &dyn Db) -> u8 {
+        match self {
+            Self::Value(value) => value,
+            Self::UntrackedRead(value) => {
+                db.report_untracked_read();
+                value
+            }
+            Self::MinIterate(inputs) => min_iterate(db, inputs),
+            Self::MaxIterate(inputs) => max_iterate(db, inputs),
+            Self::MinPanic(inputs) => min_panic(db, inputs),
+            Self::MaxPanic(inputs) => max_panic(db, inputs),
+            Self::Successor(input) => input.eval(db) + 1,
+        }
+    }
+
+    fn assert(self, db: &dyn Db, expected: u8) {
+        assert_eq!(self.eval(db), expected)
+    }
+}
+
+#[salsa::tracked(cycle_fn=min_recover, cycle_initial=min_initial)]
+fn min_iterate<'db>(db: &'db dyn Db, inputs: Inputs) -> u8 {
+    inputs.values(db).min().expect("empty inputs!")
+}
+
+const MIN_COUNT_FALLBACK: u8 = 100;
+const MIN_VALUE_FALLBACK: u8 = 5;
+const MIN_VALUE: u8 = 10;
+
+fn min_recover(_db: &dyn Db, value: &u8, count: u32) -> CycleRecoveryAction<u8> {
+    if *value < MIN_VALUE {
+        CycleRecoveryAction::Fallback(MIN_VALUE_FALLBACK)
+    } else if count > 10 {
+        CycleRecoveryAction::Fallback(MIN_COUNT_FALLBACK)
+    } else {
+        CycleRecoveryAction::Iterate
+    }
+}
+
+fn min_initial(_db: &dyn Db) -> u8 {
+    255
+}
+
+#[salsa::tracked(cycle_fn=max_recover, cycle_initial=max_initial)]
+fn max_iterate<'db>(db: &'db dyn Db, inputs: Inputs) -> u8 {
+    inputs.values(db).max().expect("empty inputs!")
+}
+
+const MAX_COUNT_FALLBACK: u8 = 200;
+const MAX_VALUE_FALLBACK: u8 = 250;
+const MAX_VALUE: u8 = 245;
+
+fn max_recover(_db: &dyn Db, value: &u8, count: u32) -> CycleRecoveryAction<u8> {
+    if *value > MAX_VALUE {
+        CycleRecoveryAction::Fallback(MAX_VALUE_FALLBACK)
+    } else if count > 10 {
+        CycleRecoveryAction::Fallback(MAX_COUNT_FALLBACK)
+    } else {
+        CycleRecoveryAction::Iterate
+    }
+}
+
+fn max_initial(_db: &dyn Db) -> u8 {
+    0
+}
+
+#[salsa::tracked]
+fn min_panic<'db>(db: &'db dyn Db, inputs: Inputs) -> u8 {
+    inputs.values(db).min().expect("empty inputs!")
+}
+
+#[salsa::tracked]
+fn max_panic<'db>(db: &'db dyn Db, inputs: Inputs) -> u8 {
+    inputs.values(db).max().expect("empty inputs!")
+}
+
+// Diagram nomenclature for nodes: Each node is represented as a:xx(ii), where `a` is a sequential
+// identifier from `a`, `b`, `c`..., xx is one of the four query kinds:
+// - `Ni` for `min_iterate`
+// - `Xi` for `max_iterate`
+// - `Np` for `min_panic`
+// - `Xp` for `max_panic`
+//
+// and `ii` is the inputs for that query, represented as a comma-separated list, with each
+// component representing an input:
+// - `a`, `b`, `c`... where the input is another node,
+// - `uXX` for `UntrackedRead(XX)`
+// - `vXX` for `Value(XX)`
+// - `sY` for `Successor(Y)`
+//
```

Comment:
These are admittedly obscure-looking chicken scratches. I'm not claiming they are super readable, but they are concise enough to put into an ASCII graph diagram, and (once you get familiar with them) give a lot of information about the behavior of the test. They were really helpful to me in writing and debugging the tests.

Open to feedback that I should do this differently for better readability by future maintainers...